js/application.js

```diff
@@ -126,194 +209,507 @@ $(document).ready(function(){
-              $('#'+gridRowPrev).children('#'+gridCol).text(sum);
-              $('#'+gridRow).children('#'+gridCol).text(0);
-              $('#'+gridRow).children('#'+gridCol).attr('myAttr','canMove');
-              $('#'+gridRow).children('#'+gridCol).attr('combine','canCombine');
-              $('#'+gridRowPrev).children('#'+gridCol).attr('combine','canCombine');
+          currentElement=$('#'+gridRow).children('#'+gridCol);
+          prevElement=$('#'+gridRowPrev).children('#'+gridCol);
+          if(currentElement.text()!="0"){
+            if (prevElement.text()=='0') {
+              sum=Number(prevElement.text())+Number(currentElement.text());
+              prevElement.text(sum);
+              currentElement.text(0);
+              currentElement.attr('myAttr','canMove');
+              currentElement.attr('combine','canCombine');
+              prevElement.attr('combine','canCombine');
               var hasMoved=true;
-            } else if (($('#'+gridRowPrev).children('#'+gridCol).text()==$('#'+gridRow).children('#'+gridCol).text())&&($('#'+gridRowPrev).children('#'+gridCol).attr('combine')=='canCombine')&&($('#'+gridRow).children('#'+gridCol).attr('combine')=='canCombine')) {
-              sum=Number($('#'+gridRowPrev).children('#'+gridCol).text())+Number($('#'+gridRow).children('#'+gridCol).text());
-              $('#'+gridRowPrev).children('#'+gridCol).text(sum);
-              $('#'+gridRow).children('#'+gridCol).text(0);
+            } else if ((currentElement.text()==prevElement.text())&&(prevElement.attr('combine')=='canCombine')&&(currentElement.attr('combine')=='canCombine')) {
+              sum=Number(prevElement.text())+Number(currentElement.text());
+              prevElement.text(sum);
               totalScore += sum;
-              $('#'+gridRow).children('#'+gridCol).attr('myAttr','canMove');
-              $('#'+gridRow).children('#'+gridCol).attr('combine','canCombine');
-              $('#'+gridRowPrev).children('#'+gridCol).attr('combine','cantCombine');
+              currentElement.text(0);
+              currentElement.attr('myAttr','canMove');
+              currentElement.attr('combine','canCombine');
+              prevElement.attr('combine','cantCombine');
               var hasMoved=true;
             } else {
-              $('#'+gridRow).children('#'+gridCol).attr('myAttr','cantMove');
+              currentElement.attr('myAttr','cantMove');
             }
           } else {
-            $('#'+gridRow).children('#'+gridCol).attr('myAttr','cantMove');
+            currentElement.attr('myAttr','cantMove');
           }
         }
       }
     }
     reset();
-    if ((hasMoved==true)&&(winState==null)){randomNumberGeneration()};
+    if ((hasMoved==true)&&((winState==null)||(over==null))) {
+      randomNumberGeneration()
+    };
     printScore();
   }
 
   var checkWin = function(){
```

Comment:
Can you just check if "2048" exists in the 2D array? So just a nested for loop would do?